smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -82,74 +164,144 @@ def process_request(self, request: DragonRequest) -> DragonResponse:
-            # stdout=Popen.PIPE,
-            # stderr=Popen.PIPE,
-        )
-
-        grp = ProcessGroup(restart=False, pmi_enabled=request.pmi_enabled)
-        grp.add_process(nproc=request.tasks, template=proc)
         step_id = self._get_new_id()
-        grp.init()
-        grp.start()
-        self._proc_groups[step_id] = (grp, grp.puids)
+        if not self._request_is_satisfiable(request):
+            message = f"Cannot satisfy request. Requested {request.nodes}, "
+            message += f"but only {len(self._hosts)} nodes are available."
+            self._group_infos[step_id] = ProcessGroupInfo(status=STATUS_FAILED)
+            return DragonRunResponse(step_id=step_id, error_message=message)
 
+        self._queued_steps[step_id] = request
+        self._group_infos[step_id] = ProcessGroupInfo(status=STATUS_NEVER_STARTED)
         return DragonRunResponse(step_id=step_id)
 
-    @process_request.register
-    def _(self, request: DragonUpdateStatusRequest) -> DragonUpdateStatusResponse:
-        updated_statuses: t.Dict[str, t.Tuple[str, t.Optional[t.List[int]]]] = {}
-        for step_id in request.step_ids:
-            return_codes: t.List[int] = []
-            if step_id in self._proc_groups:
-                proc_group_tuple = self._proc_groups[step_id]
-                if proc_group_tuple[0].status == "Running":
-                    updated_statuses[step_id] = (STATUS_RUNNING, return_codes)
-                else:
-                    if all(proc_id is not None for proc_id in proc_group_tuple[1]):
-                        try:
-                            return_codes = [
-                                Process(None, ident=puid).returncode
-                                for puid in proc_group_tuple[1]
-                            ]
-                        except (ValueError, TypeError):
-                            return_codes = [-1 for _ in proc_group_tuple[1]]
-                    else:
-                        return_codes = [0]
-                    status = (
-                        STATUS_FAILED
-                        if any(return_codes) or proc_group_tuple[0].status == "Error"
-                        else STATUS_COMPLETED
-                    )
-                    updated_statuses[step_id] = (status, return_codes)
+    def update(self) -> None:
+        self._updates += 1
+        started = []
+        for step_id, request in self._queued_steps.items():
+            hosts = self._allocate_step(step_id, self._queued_steps[step_id])
+            if not hosts:
+                continue
+
+            print(f"Step id {step_id} allocated on {hosts}")
+
+            global_policy = Policy(
+                placement=Policy.Placement.HOST_NAME, host_name=hosts[0]
+            )
+            grp = ProcessGroup(restart=False, pmi_enabled=True, policy=global_policy)
+
+            for node_num in range(request.nodes):
+                node_name = hosts[node_num]
```

Comment:
Do we need to do direct indexing here? could this just be
```suggestion
            for node_name in hosts[:request.nodes]:
```
?

Reply:
Applied

Reply:
Re-raising this request because I am too dumb to understand or think about indexing, hahahaha

Reply:
Sorry, I guess that I applied it, then I merged and lost it